### DISCUSSION TITLE
add static class api compatibility layer for pipeline .NET languages, like F#

### DISCUSSION THREAD

**Discussion by jkone27** [dotnet/aspnetcore#59661] [Ideas]
would be greate to have a way to configure .NET aspnecore by default also for F# projects,

so as an alternative of builder pattern and derivatives and fluent extensions (OO way), just pseudocode here

```csharp
new Builder()
      .WithA(.....)
      .WithB(.....)
      .Build()
      .Configure(...)
```

also provide static methods from static classes constructing new objects and returning them as result

```csharp
      var b = new Builder();
      var b2 = StaticModule.WithA(b)
      var b3 = StaticModule.WithB(b2)
      var app = Builder.Build(b3);
      return App.Configure(app);
```

so that in F# it can be used as

```fsharp
new Builder()
|> StaticModule.WithA
|> StaticModule.withB
|> Builder.Build
|> App.Configure
```

**Comment by Smaug123** [comment]
Later versions of F# let you do this (from memory; caveat emptor):

```fsharp
Builder()
|> _.WithA(...)
|> _.WithB(...)
|> _.Build()
```

This is never going to be as pretty/ergonomic as functions would be, but it's considerably better than the fluent APIs.

**Reply by pblasucci** [reply]
I routinely have `Program.fs` files that end like this:
```fsharp
[<EntryPoint>]
let main args =
  args
  |> WebApplication.CreateBuilder
  |> configureServices
  |> _.Build()
  |> configureApplication
  |> _.Run()
```
wherein `configureServices` and `configureApplication` are simple functions that do all the actual build-up. Within those functions, things can be as F#-ish or C#-ish as one desires (though, unfortunately, some extra type annotations and calls to `|> ignore` are generally needed).
